### PR TITLE
DTSRD-3529. add 2 extra columns in the LOV table

### PR DESCRIPTION
**Pull request by SabinaHMCTS**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/testsupport/CommonDataFunctionalBaseTest.java [hunk 1]
```diff
@@ -185,6 +185,30 @@ protected void validateListOfValuesFile(JdbcTemplate jdbcTemplate, String servic
         assertEquals(exceptedResult, listOfValues);
     }
 
+    protected void validateListOfValuesFile1(JdbcTemplate jdbcTemplate, String serviceSql,
```
Comment (mikebrownccd) [inline src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/testsupport/CommonDataFunctionalBaseTest.java:188]: Can this be given a more descriptive name?

File: src/functionalTest/resources/sourceFiles/categories/list_of_values_duplicate_rec_case2.csv [hunk 2]
```diff
@@ -1,3 +1,3 @@
-CategoryKey,ServiceID,Key,Value_EN,Value_CY,Hinttext_EN,Hinttext_CY,Lov_Order,ParentCategory,ParentKey,Active
-caseSubType,BBA3,BBA3-001AD,ADVANCE PAYMENT new,,,,,caseType,BBA3-001,D
-caseSubType,BBA3,BBA3-001AD,ADVANCE PAYMENT scenario2,,,,,caseType,BBA3-001,Y
+CategoryKey,ServiceID,Key,Value_EN,Value_CY,Hinttext_EN,Hinttext_CY,Lov_Order,ParentCategory,ParentKey,Active,External_Reference,External_Reference_Type
```
Comment (mikebrownccd) [inline src/functionalTest/resources/sourceFiles/categories/list_of_values_duplicate_rec_case2.csv:1]: The first line of the csv file (Headers) has 12 "," but lines 2 and 3 have 13 ","

File: src/functionalTest/resources/sourceFiles/categories/list_of_values_failure_additional_header.csv [hunk 3]
```diff
@@ -1,4 +1,4 @@
-categorykey,serviceid,key,value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active,newHeader
-AdditionalFacilities,,AF-VF,Video Facility,,,,,,,Y
-AdditionalFacilities,,AF-SSC,Same Sex Courtroom,,,,,,,Y
-AdditionalFacilities,,AF-WR,Witness Room,,,,,,,Y
+categorykey,serviceid,key,value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active,external_reference,external_reference_type,newHeader
```
Comment (mikebrownccd) [inline src/functionalTest/resources/sourceFiles/categories/list_of_values_failure_additional_header.csv:1]: 2 commas added to the header but 3 added to the lines below

File: src/main/resources/db/migration/V1_8__alter_List_Of_Values.sql [hunk 4]
```diff
@@ -0,0 +1,8 @@
+
+ALTER TABLE List_Of_Values ADD COLUMN external_reference VARCHAR(200);
+ALTER TABLE List_Of_Values ADD COLUMN external_reference_type VARCHAR(200);
+
+ALTER TABLE list_of_values
```
Comment (lukasz-wolski) [inline src/main/resources/db/migration/V1_8__alter_List_Of_Values.sql:5]: minor: readability - please decide on either lower case or capitalised case across migration file

**Review by lukasz-wolski — CHANGES_REQUESTED** [review]
please see comments inlined

File: src/main/resources/application-crd-list-of-values-router.yaml [hunk 5]
```diff
@@ -9,17 +9,18 @@ route:
...
       hinttext_cy = :#hinttext_cy, lov_order = :#lov_order, parentcategory = :#parentcategory, parentkey = :#parentkey,
-      active = :#active?batch=true&dataSource=dataSource
+      active = :#active,external_reference = :#external_reference,external_reference_type = :#external_reference_type?batch=true&dataSource=dataSource
     blob-path:
       azure-blob://${azure.storage.account-name}/rd-common-data/ListOfValues.csv?credentials=#credsreg&operation=updateBlockBlob
     processor-class: categoriesProcessor
     mapper-class: categoriesMapper
     csv-binder-object: Categories
-    csv-headers-expected: categorykey,serviceid,key,value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active
+    #csv-headers-expected: categorykey,serviceid,key,value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active
```
Comment (mikebrownccd) [inline src/main/resources/application-crd-list-of-values-router.yaml:23]: commented code

File: src/main/resources/db/migration/V1_1__init_tables.sql [hunk 6]
```diff
@@ -9,7 +9,7 @@ create table flag_details(
 	constraint id_unique unique (id)
 );
 
-INSERT INTO public.flag_details (id,flag_code,value_en,value_cy,category_id) VALUES
+INSERT INTO dbcommondata.flag_details (id,flag_code,value_en,value_cy,category_id) VALUES
```
Comment (mikebrownccd) [inline src/main/resources/db/migration/V1_1__init_tables.sql:12]: i think that flyway might not like it if we modify and already applied delta

File: src/test/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessorTest.java [hunk 7]
```diff
@@ -43,17 +45,21 @@
...
 
     private static final List<Pair<String, Long>> ZERO_BYTE_CHARACTER_RECORDS = List.of(Pair.of("BFA1-001AD", null),
-                                                                   Pair.of("BFA1-PAD", null),
-                                                                   Pair.of("BFA1-DC\u200BX", null));
+                                                                   Pair.of("BFA1-DC\u200BX", null),
+                                                                    Pair.of("BFA1-PAD", null));
+
+    public static final String LOV_EXTERNAL_REFERENCE = "external_reference,external_reference_type";
+
+    public static final String EXTERNAL_REFERENCE_ERROR_MSG = "Both external_reference_type and "
```
Comment (mikebrownccd) [inline src/test/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessorTest.java:61]: "external_reference_type" repeated in message string

File: src/test/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessorTest.java [hunk 8]
```diff
@@ -322,7 +458,7 @@ private List<Categories> getLovServicesCase5() {
                 .categoryKey("caseSubType")
                 .serviceId("BFA1")
                 .key("BFA1-001AD")
-                .valueEN("Refusal of application under the EEA regulations \u200B")
+                .valueEN("Refusal of application under the EEA regulations ")
```
Comment (mikebrownccd) [inline src/test/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessorTest.java:461]: Was this previously part of the Zero Byte test?

**Review by mikebrownccd — COMMENTED** [review]
I have left a few comments

File: src/main/resources/application-crd-list-of-values-router.yaml (hunk 5 above)
Comment (SabinaHMCTS) [reply src/main/resources/application-crd-list-of-values-router.yaml]: sorry . that's deleted now

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java [hunk 9]
```diff
@@ -38,9 +39,25 @@ public class CategoriesProcessor extends JsrValidationBaseProcessor<Categories>
...
+    private JdbcTemplate jdbcTemplate;
+
     public static final String LOV_COMPOSITE_KEY = "categorykey,key,serviceid";
+
+    public static final String LOV_EXTERNAL_REFERENCE = "external_reference,external_reference_type";
+
     public static final String LOV_COMPOSITE_KEY_ERROR_MSG = "Composite Key violation";
 
+    public static final String EXTERNAL_REFERENCE_ERROR_MSG = "Both external_reference_type and "
+        + "external_reference_type value must be null or both must be not-null";
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java:53]: Changed .

will update jira

File: bin/main/application-crd-list-of-values-router.yaml [hunk 10]
```diff
@@ -20,6 +20,6 @@ route:
     processor-class: categoriesProcessor
     mapper-class: categoriesMapper
     csv-binder-object: Categories
-    csv-headers-expected: categorykey,serviceid,key,value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active
+    csv-headers-expected: categorykey,serviceid,key,value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active,
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply bin/main/application-crd-list-of-values-router.yaml]: done

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (lukasz-wolski) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java:53]: Original wording still present. Do not see an entry in jira ticket regarding error message. Please confirm whether this is intended

**Review by lukasz-wolski — APPROVED** [review]
please see comment inlined

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java:53]: updated

[the remark this answers is not in the document]
Comment (lukasz-wolski) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java:53]: https://tools.hmcts.net/jira/browse/DTSRD-3529 does not contain any updates. Which Jira ticket are you referring to?

[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java:53]: https://tools.hmcts.net/jira/browse/DTSRD-3530

**Pull request merged**